Improve and extend entry view table

fonic opened the pull request:
## Description

This improves and extends the entry view table,
- Adding additional colums _Password_, _Notes_, _Expires_, _Created_, _Modified_, _Accessed_ and _Attachments_ (addresses #129)
- Adding settings to display usernames/passwords hidden/visible
- Adding a context menu to customize view (toggle column visibility, resize and reset columns) (addresses #17, #129, #397)
- Adding a _'copy-on-doubleclick'_ feature for certain columns (addresses #360)
- Syncing view states when switching between multiple tabs
- Reading/writing last used view settings from/to config

bringing the entry view table in par with KeePassX 0.4.x and even beyond.

**NOTE:**
- KeePassXC's user configuration should be reset prior to testing this
- This is replaces my original pull request #849, providing a clean rewrite with clean commits.

## Motivation
In my opinion, this addresses a major disadvantage of KeePassXC as the time of writing. The additional columns, the ability to customize the views and the _'copy-on-doubleclick'_ feature improve KeePassXC's overall usability and user experience.

## DONE
- Add additional columns to entry view table
- Add code for correct sorting of entries for any given column
- Add settings to display usernames/passwords hidden/visible
- Add context menu to entry view table header for view customization
- Add 'copy-on-doubleclick' feature for columns where applicable
- Add code to sync view states when switching between multiple tabs
- Add code to make view settings persistent using configuration file
- Finalize and clean up code

## Possible follow-ups
- Double-clicking columns 'Attachments' and 'Notes' should open corresponding tab in details view/pane
- Make header context menu available as 'View' in menu bar of main window
- Optionally add header context menu as submenu 'View' to already existing entry list context menu
- Copy-on-doubleclick should give feedback, e.g. progress bar that displays how long the copied element will stay in the clipboard (requested by @droidmonkey)
- Update/modify search mode to work without column 'Group' to simplify state syncing (see #1317)
- Add entropy column displaying password quality (see [this comment](https://github.com/keepassxreboot/keepassxc/pull/1305#issuecomment-354390256))

## How has this been tested?
- locally

## Screenshots
![screenshot_20171221_104349](https://user-images.githubusercontent.com/14027079/34250168-463df4f8-e63c-11e7-95c6-deb386723085.png)

## Types of changes
- ✅ New feature (non-breaking change which adds functionality)

## Checklist (in progress)
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- :negative_squared_cross_mark: My change requires a change to the documentation and I have updated it accordingly.
- :negative_squared_cross_mark: I have added tests to cover my changes.

fonic commented:
So, @droidmonkey, @TheZ3ro, @louib, this is it, done for now, just in time to be a nice Christmas present ;)

Please test.

fonic commented:
A note on on state syncing:

A spent hours trying to find ways for proper state syncing. Now I can say that, with the current/original implementation, it is not possible and never will be to achieve a clean solution. The problem is the search mode, which, by showing/hiding column 'Group', effectively taints the view state.

Thus, I'd like to suggest changing the search mode to work without this column. This would _greatly_ reduce code complexity - I'd say we're talking about 250-500 lines of code. One solution I can think of is implementing an E-Mail-client-like search, where the number of search results for each group is displayed in the group tree view.

Here's a quick mock-up:
![screenshot_20171223_092213](https://user-images.githubusercontent.com/14027079/34318219-d8794aca-e7c2-11e7-82d7-b59736d530e9.png)

Please also note that I currently crafted the state syncing to be as close to the original behavior as possible. But, as it turns out, the original state syncing _never worked_ as it was designed due to a bug in class DatabaseWidgetStateSync:
```
void DatabaseWidgetStateSync::updateColumnSizes()
{
    if (m_blockUpdates) {
        return;
    }

    if (m_activeDbWidget->isGroupSelected()) {
        m_columnSizesList = m_activeDbWidget->entryHeaderViewSizes();
    }
    else {
        m_columnSizesSearch = m_activeDbWidget->entryHeaderViewSizes();
    }
}
```
should be
```
void DatabaseWidgetStateSync::updateColumnSizes()
{
    if (m_blockUpdates) {
        return;
    }

    if (m_activeDbWidget->isInSearchMode()) {
        m_columnSizesSearch = m_activeDbWidget->entryHeaderViewSizes();
    }
    else {
        m_columnSizesList = m_activeDbWidget->entryHeaderViewSizes();
    }
}
```
It seems to me that this is either a remnant from previous implementations or a copy-and-paste error. To cut a long story short: in my implementation, this bug is fixed, which results in a somewhat different behavior.

droidmonkey commented:
I recommend an entirely different approach. Search results should appear in their own entry view with fixed columns like it does currently. The original entry view would be hidden or sorted below the stack. When search is over, the main entry view is shown again as usual. This would eliminate the need for the state sync, although you'll need to replumb some signals and slots depending on which view is visible.

My ultimate goal is to move the search display above all tabs. We should be searching across all open databases. The current implementation is a stop gap after moving the search bar to the toolbar.

fonic commented:
> I recommend an entirely different approach. Search results should appear in their own entry view with fixed columns like it does currently. The original entry view would be hidden or sorted below the stack. When search is over, the main entry view is shown again as usual. This would eliminate the need for the state sync, although you'll need to replumb some signals and slots depending on which view is visible. This widget's display abd behavior could be encapsulated into a simple "SearchResultsWidget" owned by the database widget.
> 
> My ultimate goal is to move the search display above all tabs. We should be searching across all open databases. The current implementation is a stop gap after moving the search bar to the toolbar.

Personally, I find to be the current search quite good, except for the fact that it is always global and clicking on groups when in search mode does nothing. I'd also prefer searching in only the current database. Other users might feel different. In the end, I think something like that should be configurable as you'll probably never find a solution that fits all use cases.

> I would accept your original proposal, but without adding the count to group names.
Without that and with the group column gone, you wouldn't know where exactly the listed entries are in you database. That's why I came up with the count idea in the first place.

droidmonkey commented:
The count still doesn't tell you

fonic commented:
> The count still doesn't tell you and adds complexity

I agree, but it gives you a valuable hint, and that might do the trick in certain situations. As said before, it all comes down to the use cases.

Therefore, I'd like to suggest opening an issue for this where we try to collect/construct all imaginable use cases, i.e. what would one search for and what would one expect of the search to help you reach your objective. Once we got the use cases, we can come up with ideas for implementations that would cover those.

fonic commented:
Ok, I'll create an issue for that.

> Personally, I think the current search is quite good, except for the fact that it is always global and clicking on groups when in search mode does nothing.

Just found out that this feature is actually already implemented! If one adds this to the config
```
[General]
SearchLimitGroup=true
```
the behavior is exactly as described. It seems there's currently no way to toggle this within the GUI, correct?

phoerious commented:
There is. Click on the little arrow next to the search icon.

fonic commented:
Hm, wanted to say 'what arrow?', but it seems that it just isn't visible on my system, probably related to the default KDE plasma theme (is the array dark grey?).
![screenshot_20171224_095740](https://user-images.githubusercontent.com/14027079/34325402-eba21b94-e890-11e7-8281-ac6cc8a80b77.png)

phoerious commented:
I don't quite remember if there actually is one or if we only talked about having one there. If there is one, it's not only invisible to you. ;-)

About this whole search issue: my dream search function would work like this: you type in a search term and while you type, groups are being reduced to groups with matching entries only (and their parents of course). The entry table view doesn't modify its columns, but shows all search results in the selected group and its children (you can change the scope, by clicking on a group in the left-hand pane).
Inside the table view, groups are visually separated by column-spanning rows containing the group name.

I quickly photoshopped your mockup to visualize my ideas:

![34318219-d8794aca-e7c2-11e7-82d7-b59736d530e9](https://user-images.githubusercontent.com/911270/34326086-5c690b58-e8a5-11e7-8237-abe7d9f7c2c6.png)

I think this customized table view could even be an optional standard view for groups (instead of only showing direct descendants for each group).

To tackle the problem of global search: I think that one should be optional and if we do that, a similar view should open in a new tab with a merged results view. We need to make sure to update it as soon as one of the databases is closed, though.

fonic commented:
> I don't quite remember if there actually is one or if we only talked about having one there. If there is one, it's not only invisible to you. ;-)

Thanks for clarifying!

> About this whole search issue:
> [...]
> I quickly photoshopped your mockup to visualize my ideas:
> [...]

Wow, this is great! Now that would/could be something I really like, I bet other users as well!

I created issue #1317 to further discuss this as decided above.

fonic commented:
@droidmonkey, @phoerious: could you repost your ideas in issue #1317?

I'll repost my idea there, too.

phoerious commented:
Yeah, sure.

TheZ3ro commented:
@fonic @phoerious 
> There is. Click on the little arrow next to the search icon.

It's not an arrow, it's the search icon (the magnifying glass)

That option anyway doesn't implement multi-database search, the goal @droidmonkey is trying to achieve.

phoerious commented:
> That option anyway doesn't implement multi-database search, the goal @droidmonkey is trying to achieve.

Read my last paragraph.

TheChiefMeat commented:
Can I suggest a entropy column so you can see password quality/length at a glance?

fonic commented:
> @TheChiefMeat
> Can I suggest a entropy column so you can see password quality/length at a glance?

I added this to the list of possible follow-ups in the description of this PR.

fonic commented:
Is any

fonic commented:
Closed this by accident, reopening.

fonic commented:
Close this by accident, reopening.

phoerious commented:
I'll see what I can do in the evening and over the weekend.

fonic commented:
@phoerious: no rush. It's lots of code. As I said, would just be great to finish this while the memory is still fresh.

droidmonkey commented on src/gui/entry/EntryModel.cpp:
```diff
@@ -145,17 +206,139 @@ QVariant EntryModel::data(const QModelIndex& index, int role) const
+            return result;
+        case Attachments:
+            /*
+             * Display comma-separated list of attachments
```
Instead of a list of attachments I recommend either just writing "YES"/"NO" or adding a paperclip icon.

droidmonkey replied:
ooooh good idea!

TheZ3ro replied:
I like the paperclip

fonic replied:
Ok, I'll add this column!

droidmonkey commented on src/gui/entry/EntryModel.cpp:
```diff
@@ -145,17 +206,139 @@ QVariant EntryModel::data(const QModelIndex& index, int role) const
                 result.prepend(tr("Ref: ","Reference abbreviation"));
             }
             return result;
+        case Notes:
```
I don't see the value in this, they are now displayed in the details pane below the entry list. Similar to attachments, it might just be ok to write "YES"/"NO".

droidmonkey replied:
After thinking about it, and the fact that you truncate the length, I am OK with this approach. Please disregard this comment.

fonic replied:
ok, thanks

droidmonkey commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -901,12 +938,39 @@ void DatabaseWidget::unlockDatabase(bool accepted)
+        }
+        break;
+    case EntryModel::Notes:
+        setClipboardTextAndMinimize(entry->resolveMultiplePlaceholders(entry->notes()));
```
Double clicking notes should open the details view to the notes tab instead.

fonic replied:
I'll look into it. At the time when I started implementing this in 2016, there was no details view.

droidmonkey commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -901,12 +938,39 @@ void DatabaseWidget::unlockDatabase(bool accepted)
+ *
+ * TODO:
+ * If pull request #1298 gets merged, double-clicking column 'Attachments'
+ * could open the new details view (see second screenshot)
```
This PR was merged, please implement this

fonic replied:
Now or in a follow up?

droidmonkey replied:
now would be perfect

fonic replied:
I'll have to defer this as well, as the review process has already eaten up much more time than anticipated.

droidmonkey commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -901,12 +938,39 @@ void DatabaseWidget::unlockDatabase(bool accepted)
+        setClipboardTextAndMinimize(entry->resolveMultiplePlaceholders(entry->username()));
+        break;
+    case EntryModel::Password:
+        setClipboardTextAndMinimize(entry->resolveMultiplePlaceholders(entry->password()));
```
I think we need to add a little notice to the user that the password was copied, especially if minimize to tray is not enabled. It is not obvious at all that the copy happened. Even adding a little countdown progress bar to the lower right would be awesome to also show the clipboard timeout.

This may be out of scope of this PR, but an idea that came to me while reviewing.

fonic replied:
Good idea, that's also what KeePass does.

droidmonkey replied:
agree, canceling this comment

fonic commented:
@droidmonkey:
> In general we do not put individual contributors names in the source files. Please remove all @author Fonic <https://github.com/fonic> entries in comments. You will be credited in the about dialog.

That's quite alright, I mainly put those in the code for myself, so that I can easily spot what I touched and what not. I'll remove and/or condense all the comments.

> Please rebase onto develop after the kdbx4 PR is merged.

Will do.

> Fantastic work, this really enhances the UI!

Thanks!

fonic commented:
Just as a side note: I'm already testing the integration of the 'View' menu in the menu bar, this would be part of the follow-up:

![screenshot_20180114_071144](https://user-images.githubusercontent.com/14027079/34913345-5d104084-f8fa-11e7-938a-f988fa362dca.png)

droidmonkey commented:
KDBX4 has been merged, this can be rebased.

phoerious commented on src/gui/entry/EntryModel.h:
```diff
@@ -52,13 +65,43 @@ class EntryModel : public QAbstractTableModel
 
+    /**
+     * @author Fonic <https://github.com/fonic>
+     * Slots to toggle state of 'Hide Usernames' and 'Hide Passwords' settings
```
This, too.

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -19,19 +19,43 @@
     m_sortModel->setSortCaseSensitivity(Qt::CaseInsensitive);
+    /**
+     * @author Fonic <https://github.com/fonic>
+     * Set Qt::UserRole as sort role
```
Single-line comments.

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -40,17 +64,117 @@ EntryView::EntryView(QWidget* parent)
+    /**
+     * @author Fonic <https://github.com/fonic>
+     * Create header context menu:
+     * - Actions to toggle state of 'Hide Usernames'/'Hide Passwords' settings
```
I don't think this comment is needed.

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -40,17 +64,117 @@ EntryView::EntryView(QWidget* parent)
+
+    m_columnActions = new QActionGroup(this);
+    m_columnActions->setExclusive(false);
+    for (int colidx = 1; colidx < header()->count(); colidx++) {
```
++colidx, maybe also rename to colIndex.

fonic replied:
Then lets go all the way and name it columnIndex.

phoerious replied:
Fine by me.

fonic replied:
Done

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -40,17 +64,117 @@ EntryView::EntryView(QWidget* parent)
+     * - Set default section size
+     * - Disable stretching of last section (interferes with fitting columns
+     *   to window)
+     * - Associate with context menu
```
Not really helpful either, states the obvious.

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
-void EntryView::switchToEntryListMode()
+/**
+ * @author Fonic <https://github.com/fonic>
+ * Switch to list mode, i.e. list entries of group (NOTE: previously named
```
Notes about previous naming are not needed (same below).

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
+ * Get current state of 'Hide Passwords' setting (NOTE: just pass-through for
+ * m_model)
+ */
+bool EntryView::hidePasswords() const
```
This function being const is the only thing showing that is it not actually hiding the password. I recommend renaming to isPasswordsHidden(). Rename the setter accordingly.

fonic replied:
Done

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
+    /* Sync checked state of menu actions to current state of view */
+    m_hideUsernamesAction->setChecked(m_model->hideUsernames());
+    m_hidePasswordsAction->setChecked(m_model->hidePasswords());
+    foreach (QAction *action, m_columnActions->actions()) {
```
Use C++11 range-based for loop (make sure to use asConst to cast the container to const to avoid detaching).

frostasm replied:
@fonic good article about [Qt containers and C++11 range-based loops](https://www.dvratil.cz/2015/06/qt-containers-and-c11-range-based-loops/)

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
+    m_hidePasswordsAction->setChecked(m_model->hidePasswords());
+    foreach (QAction *action, m_columnActions->actions()) {
+        if (static_cast<QMetaType::Type>(action->data().type()) != QMetaType::Int) {
+            Q_ASSERT(false);
```
Don't do Q_ASSERT(false), see beginning of this review.

fonic replied:
Done

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
+     * check here to make sure data actually IS int
+     */
+    if (static_cast<QMetaType::Type>(action->data().type()) != QMetaType::Int) {
+        Q_ASSERT(false);
```
Same here.

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
+            header()->resizeSection(colidx, header()->defaultSectionSize());
+        }
+    }
+    else {
```
return at the end of the if block, remove else.

fonic replied:
I wouldn't recommend that here, as this would unnecessarily disguise the inherent symmetry.

phoerious replied:
I wouldn't worry too much about symmetry, but more about clarity and prevention of mistakes. A return here makes it clear that we don't do anything further, end of journey. You don't need to worry about what comes next anymore. It keeps the function code short and concise and the indentation level low. That every boolean value has a negative is a trivial fact.

fonic replied:
Yeah, clarity... We're basically on the same page, but seem to have rather different views on what clarity and readability. Well, I don't have to maintain it, so I'll just change it to your liking.

fonic replied:
Done

phoerious replied:
Thanks!

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
+        if ((header()->count() - header()->hiddenSectionCount()) > 1) {
+            header()->hideSection(colidx);
+        }
+        else {
```
Ditto.

fonic replied:
Same here.

phoerious replied:
See above.

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
+     * visible columns and add remaining fraction to last visible column
+     */
+    int width = 0;
+    for (int colidx = 0; colidx < header()->count(); colidx++) {
```
++colidx

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
+        int add = avail / visible;
+        width = 0;
+        int last = 0;
+        for (int colidx = 0; colidx < header()->count(); colidx++) {
```
++colidx

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
+    }
+    int visible = header()->count() - header()->hiddenSectionCount();
+    int avail = header()->width() - width;
+    if ((visible > 0) && (avail > 0)) {
```
I think you can invert the condition and safely return here.

fonic replied:
Agreed.

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -144,24 +268,262 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
+    if (m_inSearchMode) {
+        header()->restoreState(m_defaultSearchViewState);
+    }
+    else {
```
else placement.

phoerious commented on src/gui/entry/EntryView.h:
```diff
@@ -38,31 +43,83 @@ class EntryView : public QTreeView
+
+    /**
+     * @author Fonic <https://github.com/fonic>
+     * Methods to get/set state of view
```
Move to CPP file. Generally, I also think you shouldn't state that this is a method or signal or what not. Just use a simple present-tense sentence about what it does (same as in commit messages).

phoerious commented on tests/gui/TestGui.cpp:
```diff
@@ -686,6 +686,21 @@ void TestGui::testSearch()
+ * for changed column indices due to new way of showing/hiding column Entry
+ * Model::ParentGroup. This column now has fixed index 0 wether it's shown
+ * or hidden, thus all indices need to be shifted by +1 when not in search
+ * mode (which is the case within this entire method)
```
Previous implementation was broken and has now completely been replaced. No need to keep historic comments about previously existing code. We don't maintain a public API which needs deprecation notes.

fonic commented on src/gui/DatabaseWidgetStateSync.cpp:
```diff
@@ -83,19 +113,59 @@ void DatabaseWidgetStateSync::setActive(DatabaseWidget* dbWidget)
+    if (!m_entryListViewState.isEmpty()) {
+        m_activeDbWidget->setEntryViewViewState(m_entryListViewState);
+    }
+    else {
```
Not my style, I only used this style as it was already present.

phoerious replied:
I assumed so. The old KeePassX code did it that way, but we deviated from that style, because we all agreed on its ugliness. :wink:

fonic replied:
It certainly is, along with switch-case using the same indentation (yes, I know, Qt style guide). I can change it for my additions, but that would leave a mixture of styles?

phoerious replied:
We kept case with the same indentation level, but abolished else on a separate line. Yes, it creates a mixture but a) better start now than never b) we will do a code reformatting before the 2.3 release anyway, changing the remaining occurrences of this.

fonic replied:
Done

fonic commented on src/gui/entry/EntryModel.cpp:
```diff
@@ -20,16 +20,61 @@
+ * and 'Password'
+ *
+ * TODO:
+ * Decide which of the proposed options should be used (stars, bullet, black
```
Bullets it is.

phoerious replied:
I had a quick look into how to find out what the system's using a couple of weeks/months ago, but wasn't able find a good way. If you find anything, feel free to implement it. But I'm also fine with using bullets for now.

fonic replied:
I already had a thorough look into that last year and came to the same conclusion. Even had a look at the Qt sources to check what they do for password edit widgets.

phoerious replied:
It might be a theme thing. AFAICS, Qt itself uses asterisks, but the platform theme can override it with bullets.

fonic replied:
Yes, IIRC that's exactly what takes places inside Qt sources. But it's all internal, there's no API call to determine the character that's used in the end.

phoerious replied:
Yes, bullets, please.

fonic replied:
Ok, bullets it is. Shall I change this in DetailsWidget.cpp as well?
```
m_ui->passwordLabel->setText("****");
```

phoerious replied:
Yup, go ahead.

fonic commented on src/gui/entry/EntryModel.cpp:
```diff
@@ -145,17 +220,148 @@ QVariant EntryModel::data(const QModelIndex& index, int role) const
+            }
+            return result;
         case Url:
+            //result = entry->resolveMultiplePlaceholders(entry->displayUrl());
```
Not obsolete, see TODO (same for other mentions of this).

phoerious replied:
Yeah, well, I'm generally not a friend of dead code behind comments. It kind of obfuscates the actual logic. Whenever you need to go back to a previous state, use the Git history. Adding code that's going to be there in the future, is another use-case I've rarely had. Not sure how I stand by that...

fonic replied:
That wasn't my point, so let me rephrase: we first need a decision here (expansion yes/no), so that I can remove or enable those lines.

TheZ3ro replied:
Potentially every field in `Title`, `Username`, `Password`, `Url` and `Notes` it's an "Spr"-compiled field. ("Spr" stand for "String placeholder replacement" as KeePass definition)

`attr->isReference()` will tell you if a field is a reference of another entry and so if it's a "Spr" field.

Right now the only placeholder possible for entry-fields is the reference one.

fonic replied:
Done

fonic commented on src/gui/entry/EntryModel.cpp:
```diff
@@ -145,17 +220,148 @@ QVariant EntryModel::data(const QModelIndex& index, int role) const
+        case Notes:
+            /*
+             * Display only first line of notes in simplified format like
+             * KeePassX does
```
?

phoerious replied:
I don't think we need to refer back to KeePassX. I'm not entirely certain about its future. :wink:

fonic replied:
Well, I don't know about you, but I consider KeePassX 0.4.x a great application, while the rewrite was... well... I don't think I can say this nicely. More respect for the classic ;)

phoerious replied:
The moment this gets merged, this comment loses its meaning, because then there is no difference between KeePassXC and KeePassX anymore. I'm just a little concerned about polluting the code with historic comments which will have no meaning to a developer tomorrow.

fonic commented on src/gui/entry/EntryModel.cpp:
```diff
@@ -145,17 +220,148 @@ QVariant EntryModel::data(const QModelIndex& index, int role) const
+            //result = entry->attachments()->keys().join(", ");
+
+            QList<QString> attachments = entry->attachments()->keys();
+            for (int i=0; i < attachments.size(); i++) {
```
Spaces added. ++i and i++ makes no difference in for loops.

phoerious replied:
Only if the compiler optimizes away the temporary value i++ needs to maintain compared to ++i. It's mostly a style question, but I like to be explicit about it. If you don't need the old value, use ++i, regardless of where in the code.

fonic replied:
Done

phoerious commented on src/gui/DatabaseWidget.cpp:
```diff
@@ -939,23 +927,17 @@ void DatabaseWidget::unlockDatabase(bool accepted)
-        Q_ASSERT(false);
-        return;
-    }
+    Q_ASSERT(entry);
```
Don't delete the if block. The assertion will only trigger in debug mode. A following if statement will ensure safe return (punny, isn't it?) in production.

fonic replied:
That's why I was reluctant to change this. That would mean doubling the expression, which is trivial here, but not so in EntryView.cpp. Doubling expressions is considered bad practice due to increased likeliness of errors, that's why I used Q_ASSERT(false) and most likely why that can be found at numerous places throughout the code.

Not producing a useful error message can't be an issue - you get the file name and line number, any good IDE even takes you right to the location.

phoerious replied:
Q_ASSERT(false) is pretty unhelpful, especially when using Qt which makes usage of debuggers without pretty printers a mess anyway. If the condition is simple and both lines are directly next to each other, I don't see a problem with duplicating it. It also makes sure that CI prints a useful error message, because depending on your system configuration and Qt version, you may not be able to reproduce it locally.

fonic replied:
Done

fonic commented:
Out of time for today, will add the rest to the cleanup commit tomorrow.

phoerious commented on src/gui/entry/EntryView.cpp:
```diff
@@ -269,159 +203,116 @@ Entry* EntryView::entryFromIndex(const QModelIndex& index)
-    m_hidePasswordsAction->setChecked(m_model->hidePasswords());
+    m_hideUsernamesAction->setChecked(m_model->isUsernamesHidden());
+    m_hidePasswordsAction->setChecked(m_model->isPasswordsHidden());
     foreach (QAction *action, m_columnActions->actions()) {
```
You still need to replace this with a range-base for loop.

fonic replied:
Why? foreach is recommended for Qt containers (in this case: QList< QAction* >).

phoerious replied:
No, it's not. foreach has been deprecated for a while.

fonic replied:
As I recall that was planned, but never done. As of Qt 5.10, it is still recommended, see documentation.

phoerious replied:
Nope. The prose documentation about containers still mentions it, but the QtGlobal reference lists it as deprecated since 5.7: https://doc-snapshots.qt.io/qt5-dev/qtglobal.html#Q_FOREACH

fonic replied:
Interesting, that's new to me. Can't get the range-based loop to work with asConst, though.

phoerious replied:
```
for (auto& action: asConst(m_columnActions->actions()) {

}
```
Should work. Use the asConst() implementation from core/global.h instead of the Qt-provided qAsConst() one, because that one's not available in our targeted minimum version.

fonic replied:
Already tried that and all other variants I could think of. What you suggested gives:
src/gui/entry/EntryView.cpp:294:58: error: use of deleted function ‘void asConst(const T&&) [with T = QList<QAction*>]’

TheZ3ro replied:
Hav you added `#include "core/Global.h` ?

fonic replied:
yes

phoerious replied:
I see, the move signature is disabled, so it's not accepting an rvalue. In that case, just save the result of m_columnActions->actions() in a const variable and iterate over that. Then you also don't need asConst() at all.

fonic replied:
Done

fonic commented on src/gui/entry/EntryModel.cpp:
```diff
@@ -145,17 +220,148 @@ QVariant EntryModel::data(const QModelIndex& index, int role) const
+             * Display only first line of notes in simplified format like
+             * KeePassX does
+             */
+            //result = entry->resolveMultiplePlaceholders(entry->notes().section("\n", 0, 0).simplified());
```
Same as above.

fonic commented:
This should be it.

TheZ3ro commented:
@fonic can you rebase on top of develop? Anyway, very good work :tada: 

@phoerious @droidmonkey please review your requested changes

fonic commented:
One little thing slipped through, changing to bullets in DetailsWidget.cpp. Done.

fonic commented:
Rebased.

fonic commented:
There are two things I'd like to discuss before this gets merged:

EntryModel.cpp
```
// Paperclip symbol
const QString EntryModel::PaperClipDisplay(QString("\U0001f4ce"));
```
seems to produce memory leaks. If changed to a simple string like 'x', they disappear. If fail to understand what this is about.

EntryModel.cpp
```
// String being displayed when hiding content
const QString EntryModel::HiddenContentDisplay(QString(QChar(0x25CF)).repeated(6));
```
Should we move this to some global header file, so that it can be used throughout the application, e.g. for the details pane?

TheZ3ro commented:
@fonic I would place both the paperclip and the bullet in a global file, for example in `core/Entry.cpp` or something like that.

Yes, seems a memory leak in libfontconfig AFAICS

phoerious commented:
Memory leaks in libfontconfig are very common. KeePassXC throws tons of them at exit. Nothing we can do about that.

phoerious commented:
Could you rebase your branch please, so I can merge?

fonic commented:
@TheZ3ro, @phoerious regarding libfontconfig memory leak:
Alright, thought so. I removed the related TODO from EntryModel.cpp.

@TheZ3ro:
`core/Entry.cpp` is not GUI related, should I really put it there? Well, we could also leave it as is, the paperclip symbol is only used in EntryModel.cpp, the hidden content string only in two places.

@phoerious:
Rebased!

phoerious commented:
Amazing, thanks. I think leaving the paper clip in EntryModel.cpp is fine.

fonic commented:
Ok, good. Looking forward to further contributions ;)

fonic commented:
@phoerious: you should be able to close several issues listed in the PR's description now that this got merged

jab416171 commented:
This is fantastic, thank you so much. How soon can we expect this to make it into a release?

phoerious commented:
It will be in 2.3, which is the next feature release. Not too long.

frostasm commented:
great work @fonic  👍
